presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerProxyHelper.java

```diff
@@ -0,0 +1,223 @@
+                .setMethod(httpMethod)
+                .setUri(remoteUri)
+                .setPreserveAuthorizationOnRedirect(true)
+                .setBodyGenerator(bodyGenerator);
+
+        for (String name : list(servletRequest.getHeaderNames())) {
+            if (isPrestoHeader(name) || name.equalsIgnoreCase(COOKIE)) {
+                for (String value : list(servletRequest.getHeaders(name))) {
+                    requestBuilder.addHeader(name, value);
+                }
+            }
+            else if (name.equalsIgnoreCase(USER_AGENT)) {
+                for (String value : list(servletRequest.getHeaders(name))) {
+                    requestBuilder.addHeader(name, "[Resource Manager] " + value);
+                }
+            }
+        }
+
+        StringBuilder xForwardedFor = new StringBuilder();
+        if (servletRequest.getHeader(X_FORWARDED_FOR) != null) {
+            xForwardedFor.append(servletRequest.getHeader(X_FORWARDED_FOR) + ",");
+        }
+        xForwardedFor.append(servletRequest.getRemoteAddr());
+        requestBuilder.addHeader(X_FORWARDED_FOR, xForwardedFor.toString());
+
+        return requestBuilder.build();
+    }
+
+    private static boolean isPrestoHeader(String name)
+    {
+        return name.toLowerCase(ENGLISH).startsWith("x-presto-");
+    }
+
+    private Response toResponse(ProxyResponse input)
+    {
+        Response.ResponseBuilder entity = status(input.getStatusCode()).entity(input.getBody());
+        input.getHeaders().forEach(((headerName, value) -> entity.header(headerName.toString(), value)));
+        return entity.build();
+    }
+
+    private void setupAsyncResponse(HttpServletRequest servletRequest, AsyncResponse asyncResponse, ListenableFuture<Response> future)
+    {
+        bindAsyncResponse(asyncResponse, future, executor)
+                .withTimeout(asyncTimeout, () -> status(BAD_GATEWAY)
+                        .type(TEXT_PLAIN)
+                        .entity(format("Request to remote Presto server (%s), current node (%s), timed out after %s",
+                                servletRequest.getRemoteAddr(),
+                                internalNodeManager.getCurrentNode().getNodeIdentifier(),
+                                asyncTimeout.toString()))
+                        .build());
+    }
+
+    private static class ForwardingBodyGenerator
+            implements BodyGenerator
+    {
+        private final HttpServletRequest servletRequest;
+        private final AtomicBoolean called = new AtomicBoolean();
+
+        public ForwardingBodyGenerator(HttpServletRequest servletRequest)
+        {
+            this.servletRequest = requireNonNull(servletRequest, "servletRequest is null");
+        }
+
+        @Override
+        public void write(OutputStream outputStream)
+                throws Exception
+        {
+            checkState(called.compareAndSet(false, true), "Already read servlet request body");
```

Comment:
It's important that it only be called once, as the servlet request body, once consumed, will just return empty.  If downstream code ever changes in a way for this to be called multiple times, then this will fail fast.

Reply:
That said, this shouldn't happen, so let me change it from `checkState` -> `verify`